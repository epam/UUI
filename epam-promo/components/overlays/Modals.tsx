import * as React from 'react';
import { withMods, ModalFooterCoreProps, ModalWindowProps, ModalBlockerProps, ModalHeaderCoreProps } from '@epam/uui-core';
import { ModalBlocker as uuiModalBlocker, ModalWindow as uuiModalWindow } from '@epam/uui-components';
import { FlexRow, FlexSpacer, RowMods, FlexCell } from '../layout';
import { IconButton } from '../buttons';
import { Text } from '../typography';
import { ReactComponent as CrossIcon } from '../../icons/navigation-close-24.svg';
import css from './Modals.scss';

export interface ModalBlockerMods {
    blockerShadow?: 'light' | 'dark' | 'none';
}

export const ModalBlocker = withMods<ModalBlockerProps, ModalBlockerMods>(uuiModalBlocker, mods => [
    css.modalBlocker,
    css['shadow-' + (mods.blockerShadow || 'dark')],
]);

export interface ModalWindowMods {
    width?: '300' | '420' | '600' | '900';
    height?: '300' | '700' | 'auto';
}

export const ModalWindow = withMods<ModalWindowProps, ModalWindowMods>(uuiModalWindow, mods => [
    css.modal,
    css['width-' + (mods.width || '420')],
    css['height-' + (mods.height || 'auto')],
]);

export interface ModalHeaderProps extends RowMods, ModalHeaderCoreProps {}

export const ModalHeader = React.forwardRef<HTMLDivElement, ModalHeaderProps>((props, ref) => (
    <FlexRow
        background={props.background || 'none'}
        padding={props.padding || '24'}
        vPadding="12"
        borderBottom={props.borderBottom ? 'gray40' : undefined}
        cx={[css.modalHeader, props.cx]}
        spacing="12"
        ref={ref}
        rawProps={props.rawProps}
    >
        {props.title && (
            <Text size="48" fontSize="18" font="sans-semibold">
                {props.title}
            </Text>
        )}
        {props.children}
        {props.onClose && <FlexSpacer />}
        {props.onClose && (
            <FlexCell shrink={0} width="auto">
                <IconButton icon={CrossIcon} onClick={props.onClose} />
            </FlexCell>
        )}
    </FlexRow>
));

export interface ModalFooterProps extends RowMods, ModalFooterCoreProps {}

export const ModalFooter = React.forwardRef<HTMLDivElement, ModalFooterProps>((props, ref) => (
    <FlexRow
        ref={ref}
        spacing={props.spacing || '12'}
        cx={[css.modalFooter, props.borderTop && css.borderTop, props.cx]}
        padding={props.padding || '24'}
        vPadding={props.vPadding || '24'}
        background={props.background || 'none'}
        rawProps={props.rawProps}
    >
        {props.children}
    </FlexRow>
));
