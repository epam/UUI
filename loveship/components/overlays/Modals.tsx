import React from 'react';
import { withMods, ModalFooterCoreProps, ModalWindowProps, ModalHeaderCoreProps, ModalBlockerProps } from '@epam/uui-core';
import { ModalBlocker as uuiModalBlocker, ModalWindow as uuiModalWindow } from '@epam/uui-components';
import { FlexRow, FlexSpacer, RowMods, FlexCell } from '../layout';
import { IconButton } from '../buttons';
import { Text } from '../typography';
import { ReactComponent as CrossIcon } from '../icons/navigation-close-24.svg';
import css from './Modals.scss';

export interface ModalBlockerMods {
    blockerShadow?: 'light' | 'dark' | 'none';
}

export const ModalBlocker = withMods<ModalBlockerProps, ModalBlockerMods>(uuiModalBlocker, mods => [
    css.modalBlocker,
    css['shadow-' + (mods.blockerShadow || 'none')],
]);

export interface ModalWindowMods {
    width?: '300' | '420' | '480' | '600' | '900';
    height?: '300' | '700' | 'auto';
}

export const ModalWindow = withMods<ModalWindowProps, ModalWindowMods>(uuiModalWindow, mods => [
    css.modal,
    css['width-' + (mods.width || '420')],
    css['height-' + (mods.height || 'auto')],
]);

export interface ModalHeaderProps extends RowMods, ModalHeaderCoreProps {}

export const ModalHeader = React.forwardRef<HTMLDivElement, ModalHeaderProps>((props, ref) => (
    <FlexRow
        background={props.background || 'none'}
        padding={props.padding || '24'}
        vPadding="12"
        ref={ref}
        borderBottom={props.borderBottom ? 'night400' : undefined}
        cx={[css.modalHeader, props.cx]}
        rawProps={props.rawProps}
    >
        {props.title && (
            <Text size="48" fontSize="18" font="sans-semibold">
                {props.title}
            </Text>
        )}
        {props.children}
        <FlexSpacer />
        {props.onClose && (
            <FlexCell shrink={0} width="auto">
                <IconButton icon={CrossIcon} onClick={props.onClose} />
            </FlexCell>
        )}
    </FlexRow>
));

export interface ModalFooterProps extends RowMods, ModalFooterCoreProps {}

export const ModalFooter = React.forwardRef<HTMLDivElement, ModalFooterProps>((props, ref) => (
    <FlexRow
        spacing="12"
        cx={[css.modalFooter, props.borderTop && css.borderTop, props.cx]}
        padding={props.padding || '24'}
        vPadding="24"
        ref={ref}
        background={props.background || 'none'}
        rawProps={props.rawProps}
        children={props.children}
    />
));
