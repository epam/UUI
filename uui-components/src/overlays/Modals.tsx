import * as React from 'react';
import { UuiContexts, UuiContext } from '@epam/uui-core';

export class Modals extends React.Component {
    static contextType = UuiContext;
    context: UuiContexts;

    constructor(props: {}) {
        super(props);
    }

    componentDidMount() {
        if (!this.context) return;
        this.context.uuiModals.subscribe(() => this.forceUpdate());
    }

    public render() {
        return (
            <>
                {this.context.uuiModals.getOperations().map(modalOperation => {
                    return React.createElement(modalOperation.component, modalOperation.props);
                })}
            </>
        );
    }
}
