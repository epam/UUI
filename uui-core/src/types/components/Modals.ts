import * as React from 'react';
import { IHasChildren, IHasCX, IHasRawProps, IModal, VPanelProps, IHasForwardedRef } from '../props';

export interface ModalWindowProps extends VPanelProps, IHasForwardedRef<HTMLDivElement> {}

export interface ModalBlockerProps
    extends IModal<any>,
        IHasCX,
        IHasChildren,
        IHasRawProps<React.HTMLAttributes<HTMLDivElement>>,
        IHasForwardedRef<HTMLDivElement> {
    disableFocusLock?: boolean;
    disableCloseByEsc?: boolean;
    disallowClickOutside?: boolean;
}

export interface ModalHeaderCoreProps extends IHasChildren, IHasCX, IHasRawProps<React.HTMLAttributes<HTMLDivElement>> {
    onClose?: () => any;
    title?: React.ReactNode;
}

export interface ModalFooterCoreProps extends IHasChildren, IHasCX, IHasRawProps<React.HTMLAttributes<HTMLDivElement>> {
    borderTop?: boolean;
}
