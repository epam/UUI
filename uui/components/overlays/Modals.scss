@use  '../../assets/styles/index' as *;

.modal-blocker {
    animation-duration: 0.3s;
    animation-name: animateModalBlocker;
    opacity: 1;
    transition: all 0.3s ease-out;

    @keyframes animateModalBlocker {
        from { opacity: 0; }
        to { opacity: 1; }
    }

    &.blocker-overlay {
        :global(.uui-modal-blocker) {
            background-color: var(--uui-modals-overlay);
        }
    }
}

.modal {
    &:global(.uui-modal-window) {
        background-color: var(--uui-modals-bg);
        overflow: auto;
        border-radius: var(--uui-modals-border-radius);
        max-height: 80%;
        max-width: 100vw;
        box-shadow: var(--uui-modals-shadow);

        @media (max-width: 640px) {
            min-width: 100vw;
            max-height: 100vh;
            border-radius: 0;
        }

        :global(.uui-shadow-top)::before {
            background: var(--uui-divider);
            height: 1px;
        }

        :global(.uui-shadow-bottom)::after {
            background: var(--uui-divider);
            height: 1px;
        }
    }
}

.modal-footer {
    border-bottom-left-radius: var(--uui-modals-border-radius);
    border-bottom-right-radius: var(--uui-modals-border-radius);

    &.border-top {
        border-top: 1px solid var(--uui-divider);
    }
}

.modal-header {
    border-top-left-radius: var(--uui-modals-border-radius);
    border-top-right-radius: var(--uui-modals-border-radius);
}

