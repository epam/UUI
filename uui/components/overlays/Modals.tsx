import * as React from 'react';
import { withMods, ModalFooterCoreProps, ModalWindowProps as uuiModalWindowProps, ModalBlockerProps, ModalHeaderCoreProps } from '@epam/uui-core';
import { ModalBlocker as uuiModalBlocker, ModalWindow as uuiModalWindow } from '@epam/uui-components';
import { FlexRow, FlexSpacer, RowMods, FlexCell, FlexRowProps } from '../layout';
import { IconButton } from '../buttons';
import { Text } from '../typography';
import { isMobile } from '@epam/uui-core';
import { ReactComponent as CrossIcon } from '@epam/assets/icons/navigation-close-outline.svg';
import css from './Modals.module.scss';

export const ModalBlocker = /* @__PURE__ */withMods<ModalBlockerProps>(uuiModalBlocker, () => [css.modalBlocker]);

interface ModalWindowMods {
    /**
     * Defines component width.
     * @default '420px'
     */
    width?: number | string;
    /**
     * Defines component height.
     * @default 'auto'
     */
    height?: number | string;
    /**
     * Defines component max-height.
     * @default '80vh'
     */
    maxHeight?: number | string;
}

export interface ModalWindowCoreProps extends uuiModalWindowProps {}

export interface ModalWindowProps extends ModalWindowCoreProps, ModalWindowMods {}

export const ModalWindow = /* @__PURE__ */withMods<uuiModalWindowProps, ModalWindowMods>(
    uuiModalWindow,
    () => [css.root, css.modal],
    (props) => {
        const normalize = (d: number | string | undefined): string | undefined => {
            if (typeof d === 'number') {
                return `${d}px`;
            }
            return d;
        };
        const width = normalize(props.width) || '420px';
        const height = normalize(props.height) || 'auto';
        const maxHeight = isMobile() ? '100dvh' : (normalize(props.maxHeight) || '80dvh');
        return {
            style: {
                ...props.style,
                width,
                height,
                maxHeight,
            },
        };
    },
);

export interface ModalHeaderProps extends RowMods, ModalHeaderCoreProps {}

export class ModalHeader extends React.Component<ModalHeaderProps> {
    render() {
        return (
            <FlexRow
                padding={ this.props.padding || '24' }
                vPadding="12"
                borderBottom={ this.props.borderBottom }
                cx={ [css.root, css.modalHeader, this.props.cx] }
                columnGap="12"
                rawProps={ this.props.rawProps }
            >
                {this.props.title && (
                    <Text size="48" fontSize="18" fontWeight="600">
                        {this.props.title}
                    </Text>
                )}
                {this.props.children}
                {this.props.onClose && <FlexSpacer />}
                {this.props.onClose && (
                    <FlexCell shrink={ 0 } width="auto">
                        <IconButton rawProps={ { 'aria-label': 'Close modal' } } icon={ CrossIcon } onClick={ this.props.onClose } />
                    </FlexCell>
                )}
            </FlexRow>
        );
    }
}

export interface ModalFooterProps extends FlexRowProps, ModalFooterCoreProps {}

export class ModalFooter extends React.Component<ModalFooterProps> {
    render() {
        return (
            <FlexRow
                columnGap={ this.props.columnGap || '12' }
                cx={ [
                    css.root,
                    css.modalFooter,
                    this.props.cx,
                ] }
                borderTop={ this.props.borderTop }
                padding={ this.props.padding || '24' }
                vPadding={ this.props.vPadding || '24' }
                rawProps={ this.props.rawProps }
            >
                {this.props.children}
            </FlexRow>
        );
    }
}
