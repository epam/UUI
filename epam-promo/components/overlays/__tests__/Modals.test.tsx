import React from 'react';
import { ModalBlocker, ModalHeader, ModalFooter, ModalWindow } from '../Modals';
import renderer from 'react-test-renderer';

describe('Modals', () => {
    it('should be rendered correctly', () => {
        const tree = renderer
            .create(
                <ModalBlocker key="blocker" isActive zIndex={1} abort={jest.fn} success={jest.fn}>
                    <ModalWindow>
                        <ModalHeader />
                        <ModalFooter />
                    </ModalWindow>
                </ModalBlocker>
            )
            .toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('should be rendered correctly', () => {
        const tree = renderer
            .create(
                <ModalBlocker key="blocker" isActive zIndex={1} abort={jest.fn} success={jest.fn} blockerShadow="dark" disallowClickOutside>
                    <ModalWindow height="300" width="300" onClick={jest.fn}>
                        <ModalHeader
                            title="Test header"
                            onClose={jest.fn}
                            borderBottom
                            margin="12"
                            background="gray5"
                            size="36"
                            spacing="6"
                            padding="6"
                            topShadow
                            vPadding="12"
                        />
                        <ModalFooter borderBottom margin="12" background="white" size="48" spacing="12" padding="18" topShadow vPadding="24" />
                    </ModalWindow>
                </ModalBlocker>
            )
            .toJSON();
        expect(tree).toMatchSnapshot();
    });
});
