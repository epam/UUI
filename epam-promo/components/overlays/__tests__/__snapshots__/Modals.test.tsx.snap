// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modals should be rendered correctly 1`] = `
<div
  className="container modalBlocker shadow-dark"
  style={
    Object {
      "zIndex": 1,
    }
  }
>
  <div
    className="uui-modal-blocker"
    onClick={[Function]}
  />
  <div
    data-focus-guard={true}
    style={
      Object {
        "height": "0px",
        "left": "1px",
        "overflow": "hidden",
        "padding": 0,
        "position": "fixed",
        "top": "1px",
        "width": "1px",
      }
    }
    tabIndex={0}
  />
  <div
    data-focus-guard={true}
    style={
      Object {
        "height": "0px",
        "left": "1px",
        "overflow": "hidden",
        "padding": 0,
        "position": "fixed",
        "top": "1px",
        "width": "1px",
      }
    }
    tabIndex={1}
  />
  <div
    data-focus-lock-disabled={false}
    onBlur={[Function]}
    onFocus={[Function]}
  >
    <div
      aria-modal={true}
      className="uui-modal-window modal width-420 height-auto container"
      role="modal"
    >
      <div
        className="root size-36 background-none padding-24 vPadding-12 spacing-12 modalHeader container align-items-center"
      />
      <div
        className="root size-36 background-none padding-24 vPadding-24 spacing-12 modalFooter container align-items-center"
      />
    </div>
  </div>
  <div
    data-focus-guard={true}
    style={
      Object {
        "height": "0px",
        "left": "1px",
        "overflow": "hidden",
        "padding": 0,
        "position": "fixed",
        "top": "1px",
        "width": "1px",
      }
    }
    tabIndex={0}
  />
</div>
`;

exports[`Modals should be rendered correctly 2`] = `
<div
  className="container modalBlocker shadow-dark"
  style={
    Object {
      "zIndex": 1,
    }
  }
>
  <div
    className="uui-modal-blocker"
    onClick={[Function]}
  />
  <div
    data-focus-guard={true}
    style={
      Object {
        "height": "0px",
        "left": "1px",
        "overflow": "hidden",
        "padding": 0,
        "position": "fixed",
        "top": "1px",
        "width": "1px",
      }
    }
    tabIndex={0}
  />
  <div
    data-focus-guard={true}
    style={
      Object {
        "height": "0px",
        "left": "1px",
        "overflow": "hidden",
        "padding": 0,
        "position": "fixed",
        "top": "1px",
        "width": "1px",
      }
    }
    tabIndex={1}
  />
  <div
    data-focus-lock-disabled={false}
    onBlur={[Function]}
    onFocus={[Function]}
  >
    <div
      aria-modal={true}
      className="uui-modal-window modal width-300 height-300 container"
      role="modal"
    >
      <div
        className="root size-36 background-gray5 padding-6 vPadding-12 border-bottom-gray40 spacing-12 modalHeader container align-items-center"
      >
        <div
          className="root font-sans-semibold text-color-gray80 line-height-24 font-size-18 v-padding-12 container"
        >
          Test header
        </div>
        <div
          className="flexSpacer"
        />
        <div
          className="flexCell"
          onClick={[Function]}
          style={
            Object {
              "alignSelf": undefined,
              "flexBasis": "auto",
              "flexGrow": undefined,
              "flexShrink": 0,
              "minWidth": 0,
              "textAlign": undefined,
            }
          }
        >
          <div
            className="root icon-color-gray60 uui-button-box uui-enabled -clickable"
            href={null}
            onClick={[Function]}
            onKeyDown={[Function]}
            role="button"
            tabIndex={0}
          >
            <div
              className="container uui-icon uui-enabled"
              style={Object {}}
            >
              <svg
                className=""
                height={24}
                viewBox="0 0 12 24"
                width={12}
              >
                <use
                  xlinkHref="#navigation-close-24.svg"
                />
              </svg>
            </div>
          </div>
        </div>
      </div>
      <div
        className="root size-36 background-white padding-18 vPadding-24 spacing-12 modalFooter container align-items-center"
      />
    </div>
  </div>
  <div
    data-focus-guard={true}
    style={
      Object {
        "height": "0px",
        "left": "1px",
        "overflow": "hidden",
        "padding": 0,
        "position": "fixed",
        "top": "1px",
        "width": "1px",
      }
    }
    tabIndex={0}
  />
</div>
`;
